torch/nn/_functions/thnn/sparse.py

```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
+        self._indices = None
+        assert mode is not None
+        if mode == 'sum':
+            self.mode = 0
+        elif mode == 'mean':
+            self.mode = 1
+        else:
+            raise InputError("mode needs to be 'sum' or 'mean', but got {}" % mode)
+
+    def _renorm(self, indices, weight):
+        self._backend.LookupTable_renorm(
+            self._backend.library_state,
+            indices,
+            weight,
+            self.max_norm,
+            self.norm_type
+        )
+
+    def forward(self, indices, offsets, weight):
+        assert indices.dim() <= 2
+        assert not self.needs_input_grad[0], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the indices"
+
+        assert not self.needs_input_grad[1], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the offsets"
+
+        self._backend = type2backend[type(weight)]
+        self._weight_size = weight.size()
+        self._offset2bag = offsets.new()
+
+        if not indices.is_contiguous():
+            self._indices = indices.contiguous()
+            indices = self._indices
+        else:
+            self.save_for_backward(indices)
+
+        output = weight.new()
+        if self.max_norm is not None:
+            self._renorm(indices, weight)
+
+        assert indices.dim() == 1
+        assert offsets.dim() == 1
+        assert offsets[0] == 0
+        assert offsets[-1] < indices.size(0)
+
+        if weight.is_cuda:
+            if self.mode == 1:
+                self.sequence_length = offsets.new().resize_(offsets.size())
+            else:
+                self.sequence_length = None
+
+            self._backend.LookupTableBag_updateOutput(
+                self._backend.library_state,
+                indices,
+                offsets,
+                weight,
+                output,
+                self._offset2bag,
+                self.mode,
+                self.sequence_length
+            )
+        else:
+            # slow CPU implementation
+            index_output = torch.index_select(weight, 0, indices)
+            # indices = [1, 2, 30, 100, 12], offsets = [0, 2, 3]
+            self._offset2bag.resize_(indices.size(0)).zero_()  # offset2bag = [0 0 0 0 0]
+            self._offset2bag.index_fill_(0, offsets, 1)  # offset2bag = [1 0 1 0 1]
+            self._offset2bag[0] = 0  # offset2bag = [0 0 1 0 1]
+            self._offset2bag = self._offset2bag.cumsum(0)  # offset2bag = [0 0 1 1 2]
+            output.resize_(offsets.size(0), weight.size(1)).zero_()
+            output.index_add_(0, self._offset2bag, index_output)
+            if self.mode == 1:
+                self.sequence_length = weight.new().resize_(offsets.size())
+                self.sequence_length[:-1] = offsets[1:] - offsets[:-1]
+                self.sequence_length[-1] = indices.size(0) - offsets[-1]
+                self.sequence_length = self.sequence_length[:, None].expand_as(output)
```

Comment:
is `sequence_length` the same as a vector of bag sizes? maybe it's a better name for it